forge/src/runner.rs

```diff
@@ -369,6 +437,147 @@ impl<'a> ContractRunner<'a> {
         })
     }
 
+    #[tracing::instrument(name = "invariant-test", skip_all)]
+    pub fn run_invariant_test(
+        &mut self,
+        runner: TestRunner,
+        setup: TestSetup,
+        test_options: TestOptions,
+        functions: Vec<&Function>,
+        known_contracts: Option<&BTreeMap<ArtifactId, (Abi, Vec<u8>)>>,
+        identified_contracts: BTreeMap<Address, (String, Abi)>,
+    ) -> Result<Vec<TestResult>> {
+        let empty = BTreeMap::new();
+        let project_contracts = known_contracts.unwrap_or(&empty);
+        let TestSetup { address, logs, traces, labeled_addresses, .. } = setup;
+
+        let start = Instant::now();
+        let prev_db = self.executor.backend().db.clone();
+        let mut evm = InvariantExecutor::new(
+            &mut self.executor,
+            runner,
+            &identified_contracts,
+            project_contracts,
+        );
+
+        let invariant_contract =
+            InvariantContract { address, invariant_functions: functions, abi: self.contract };
+
+        if let Some(InvariantFuzzTestResult { invariants, cases, reverts }) = evm.invariant_fuzz(
+            invariant_contract,
+            InvariantTestOptions {
+                depth: test_options.invariant_depth,
+                fail_on_revert: test_options.invariant_fail_on_revert,
+                call_override: test_options.invariant_call_override,
+            },
+        )? {
+            let results = invariants
+                .iter()
+                .map(|(_k, test_error)| {
```

Comment:
I'm not a fan of this large iterator here because it also contains additional levels which makes the thing a bit hard to read

perhaps this can be structured a bit more clearly